From: John Wright <jsw@debian.org>
Date: Mon, 18 Apr 2011 02:23:53 -0700
Subject: Build both static and dynamic binaries

Signed-off-by: John Wright <jsw@debian.org>
---
 Makefile |    9 +++++++--
 1 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 4b4fef2..546de2f 100644
--- a/Makefile
+++ b/Makefile
@@ -26,13 +26,16 @@ SRC	= makedumpfile.c makedumpfile.h diskdump_mod.h
 SRC_ARCH = arm.c x86.c x86_64.c ia64.c ppc64.c s390x.c
 OBJ_ARCH = arm.o x86.o x86_64.o ia64.o ppc64.o s390x.o
 
-all: makedumpfile
+all: makedumpfile makedumpfile-static
 
 $(OBJ_ARCH): $(SRC_ARCH)
 	$(CC) $(CFLAGS_ARCH) -c -o ./$@ ./$(@:.o=.c) 
 
-makedumpfile: $(SRC) $(OBJ_ARCH)
+makedumpfile-static: $(SRC) $(OBJ_ARCH)
 	$(CC) $(CFLAGS) $(OBJ_ARCH) -o $@ $< -static -ldw -lelf -lz
+
+makedumpfile: $(SRC) $(OBJ_ARCH)
+	$(CC) $(CFLAGS) $(OBJ_ARCH) -o $@ $< -ldw -lelf -lz
 	echo .TH MAKEDUMPFILE 8 \"$(DATE)\" \"makedumpfile v$(VERSION)\" \"Linux System Administrator\'s Manual\" > temp.8
 	grep -v "^.TH MAKEDUMPFILE 8" makedumpfile.8 >> temp.8
 	mv temp.8 makedumpfile.8
@@ -40,9 +43,11 @@ makedumpfile: $(SRC) $(OBJ_ARCH)
 
 clean:
 	rm -f $(OBJ) $(OBJ_ARCH) makedumpfile makedumpfile.8.gz
+	rm -f makedumpfile-static
 
 install:
 	install -D -m755 ./makedumpfile ${DESTDIR}/usr/bin/makedumpfile
+	install -D -m755 ./makedumpfile-static ${DESTDIR}/bin/makedumpfile-static
 	install -D -m755 ./makedumpfile-R.pl ${DESTDIR}/usr/bin/makedumpfile-R.pl
 	install -D -m755 ./makedumpfile.8 ${DESTDIR}/usr/share/man/man8/makedumpfile.8
 
-- 
